### ISSUE TITLE
Spurious warning about waiting in non-leaf task

### ISSUE THREAD

**Issue by manopapad** [StanfordLegion/legion#457]
If you run this program:
```
import 'regent'
task foo(x : int)
  return x + 1
end
task main()
  var x = foo(0)
  var y = foo(x)
end
regentlib.start(main)
```
with `-lg:warn`, you get this warning:
```
[0 - 7ff446e32700] {4}{runtime}: [warning 1000] LEGION WARNING: Waiting on a future in non-leaf task main (UID 1) is a violation of Legion's deferred execution model best practices. You may notice a severe performance degradation. (from file /home/manolis/legion/runtime/legion/runtime.cc:371)[warning 1000] 
```
It seems like a future is created to hold the result of the first invocation of `foo`, but the `main` task doesn't actually block on that future, but passes it to the second invocation.

**Comment by streichler** [comment]
@manopapad is this still a problem?  If so, we need to assign it to somebody...

**Comment by manopapad** [comment]
This is still happening. I know for a fact that the main function in this program doesn't block on the first invocation of `foo`; just add some debug output:
```
import 'regent'
local C = terralib.includec('unistd.h')
task foo(x : int)
  C.sleep(3)
  regentlib.c.printf('foo done\n')
  return x + 1
end
task main()
  var x = foo(0)
  var y = foo(x)
  regentlib.c.printf('main done\n')
end
regentlib.start(main)
```
and you will get this output:
```
main done
foo done
[0 - 7fa86606b700] {4}{runtime}: [warning 1000] LEGION WARNING: Waiting on a future in non-leaf task main (UID 1) is a violation of Legion's deferred execution model best practices. You may notice a severe performance degradation. (from file /home/manolis/sw/legion/runtime/legion/runtime.cc:370)[warning 1000] 
For more information see:
http://legion.stanford.edu/messages/warning_code.html#warning_code_1000

foo done
```
You can see that `main` has already exited before the first invocation of `foo` has completed, and before the warning is printed. @elliottslaughter and/or @lightsighter can comment on what might be causing this, and pick this up accordingly.

**Comment by lightsighter** [comment]
Why is this a runtime bug? From my perspective anytime you wait on a future that is a blocking call and that is a bad thing for performance. At the point in the program where you do the wait, the runtime has no context to know about whether this future was user-made or not and whether it should have triggered or not (e.g. was it passed as a task argument). The assumption is that a higher-level user (either the programmer or the compiler in this case) will be able to properly put in the arguments to silence the warnings where it safe to do so. It seems unreasonable to expect the runtime to be able to do this.

**Comment by lightsighter** [comment]
Should be fixed with 682890fec Assigning back to @manopapad to confirm.

**Comment by manopapad** [comment]
Fixed
